components/VoiceMessageSection/VoiceMessageSection.vue

```diff
@@ -0,0 +1,251 @@
+            {{ localI18n['voice-message.blockedWarning'] }}
+          </div>
+          <span
+            v-if="triedToSend && !messagePackage.audioMessage"
+            class="voice-message__warning"
+          >
+            {{ localI18n['voice-message.error.audioMessage'] }}
+          </span>
+        </div>
+        <Button class="voice-message__button" kind="primary" type="submit" :disabled="isRecording || messageSent">
+          {{ localI18n['voice-message.send-btn'] }}
+        </Button>
+      </form>
+      <div class="voice-message__notify">
+        {{ uploadStatus }}
+      </div>
+    </div>
+  </section>
+</template>
+
+<script>
+import MicRecorder from 'mic-recorder-to-mp3'
+import localI18n from '../../data/resources/i18n.json'
+import { uploadFile, saveData } from '../../services/firebase'
+import Button from '~/components/Button/Button.vue'
+export default {
+  name: 'VoiceMessageSection',
+  components: {
+    Button
+  },
+  data() {
+    return {
+      rec: null,
+      isRecording: false,
+      messageChunks: [],
+      audioPlayback: null,
+      recordingTimer: null,
+      blockedWarning: false,
+      triedToSend: false,
+      stream: null,
+      maxRecordTime: 30000,
+      uploadStatus: null,
+      messagePackage: {
+        parish: '',
+        street: '',
+        date: null,
+        audioMessage: null
+      },
+      messageSent: false,
+      localI18n
+    }
+  },
+  mounted() {
+    if (window)
+      window.AudioContext = window.AudioContext || window.webkitAudioContext
+  },
+  methods: {
+    startRecording() {
+      this.rec = new MicRecorder({ bitRate: 128 })
+
+      try {
+        this.rec.start().catch(() => {})
+
+        this.recordingTimer = setTimeout(
+          this.stopRecording.bind(this),
+          this.maxRecordTime
+        )
+        this.blockedWarning = false
+        this.isRecording = true
+      } catch (error) {
+        if (error.name === 'NotAllowedError') {
+          this.blockedWarning = true
+        }
+      }
+    },
+    stopRecording() {
+      clearTimeout(this.recordingTimer)
+
+      this.isRecording = false
+      this.rec
+        .stop()
+        .getMp3()
+        .then(([buffer, blob]) => {
+          // do what ever you want with buffer and blob
+          // Example: Create a mp3 file and play
+          const file = new File(buffer, 'voice.mp3', {
+            type: blob.type,
+            lastModified: Date.now()
+          })
+          this.messagePackage.audioMessage = file
+          this.audioPlayback = URL.createObjectURL(file)
+        })
+        .catch(() => {})
+    },
+    async checkForm(e) {
+      e.preventDefault()
+
+      this.triedToSend = true
+
+      if (this.messagePackage.date) {
+        const matches = /^(\d{4})[-\/](\d{1,2})[-\/](\d{1,2})$/.exec(
+          this.messagePackage.date
+        )
+
+        if (matches == null) {
+          this.messagePackage.date = null
+          return false
+        }
```

Comment:
should we move part of this to external functions? Like construct name and things like that

Reply:
we should. i just didn't have the time yet